src/python/bot/tasks/testcase_manager.py

```diff
@@ -526,121 +548,144 @@ def test_for_crash_with_retries(testcase,
+      # crash.
+      if not expected_state:
+        expected_state = state.crash_state
 
-  round_number = 0
-  for round_number in range(1, crash_retries + 1):
-    # Bail out early if there is no hope of finding a reproducible crash.
-    if (crash_retries - round_number + crash_count + 1 <
-        reproducible_crash_target_count):
-      break
+      if crash_result.is_security_issue() != expected_security_flag:
+        logs.log('Detected a crash without the correct security flag.')
+        continue
 
-    run_timeout = warmup_timeout if round_number == 1 else test_timeout
-    return_code, crash_time, output = process_handler.run_process(
-        command,
-        timeout=run_timeout,
-        gestures=gestures,
-        current_working_directory=app_directory)
-    process_handler.terminate_stale_application_instances()
+      crash_comparer = CrashComparer(state.crash_state, expected_state)
+      if not crash_comparer.is_similar():
+        logs.log(
+            'Detected a crash with an unrelated state: '
+            'Expected(%s), Found(%s).' % (expected_state, state.crash_state))
+        continue
 
-    crash_result = CrashResult(return_code, crash_time, output)
-    if not crash_result.is_crash():
-      logs.log(
-          'No crash occurred (round {round_number}).'.format(
-              round_number=round_number),
-          output=output)
-      continue
+      crash_count += 1
+      if crash_count >= reproducible_crash_target_count:
+        logs.log('Crash is reproducible.')
+        return True
 
-    state = crash_result.get_symbolized_data()
-    logs.log(
-        ('Crash occurred in {crash_time} seconds (round {round_number}). '
-         'State:\n{crash_state}').format(
-             crash_time=crash_time,
-             round_number=round_number,
-             crash_state=state.crash_state),
-        output=state.crash_stacktrace)
+    logs.log('Crash is not reproducible. Crash count: %d/%d.' % (crash_count,
+                                                                 round_number))
+    return False
 
-    # If we don't have an expected crash state, set it to the one from initial
-    # crash.
-    if not expected_state:
-      expected_state = state.crash_state
 
-    if crash_result.is_security_issue() != expected_security_flag:
-      logs.log('Detected a crash without the correct security flag.')
-      continue
+def test_for_crash_with_retries(testcase,
+                                testcase_path,
+                                test_timeout,
+                                http_flag=False,
+                                compare_crash=True):
+  """Test for a crash and return crash parameters like crash type, crash state,
+  crash stacktrace, etc."""
+  runner = TestcaseRunner(testcase.fuzzer_name, testcase_path, test_timeout,
+                          testcase.gestures, http_flag)
+  crash_retries = environment.get_value('CRASH_RETRIES')
```

Comment:
can we just initialize crash_retries in TestcaseRunner itself, i don't see we are setting anything custom in test_for_crash_with_retries or test_for_reproducibility

Reply:
I did this to make TestcaseRunner more general and not depend on more global state (i.e. an environment variable).

Reply:
Sounds good, thanks!